File: packages/go/singleton.go
package blasphem

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"
)

var (
	singleton    sync.RWMutex
	current      *Instance
	currentKey   string
	initializing sync.Mutex
)

func optionsKey(options Options) string {
	locales := make([]string, 0, len(options.Locales))
	for _, locale := range options.Locales {
		locales = append(locales, fmt.Sprintf("%s:%x:%s:%x:%s",
			strings.ToLower(strings.TrimSpace(locale.Code)),
			sha256.Sum256(locale.Pack), locale.PackSHA256,
			sha256.Sum256(locale.Detect), locale.DetectSHA256))
	}
	for _, code := range options.LocaleCodes {
		locales = append(locales, strings.ToLower(strings.TrimSpace(code)))
	}
	sort.Strings(locales)
	key, _ := json.Marshal(struct {
		Locales  []string
		Assets   string
		HasPacks bool
		NoDetect bool
		Grawlix  bool
	}{locales, options.Assets, options.Packs != nil, options.DisableDetection, options.Grawlix})
	return string(key)
}

// Init loads the locales and installs the package judge. The same options
// again reuse it. Different options build a new judge first and retire the
// old one after, so Judge has no gap. A failed Init keeps the previous judge.
func Init(options Options) error {
	initializing.Lock()
	defer initializing.Unlock()
	key := optionsKey(options)
	singleton.RLock()
	same := current != nil && currentKey == key && options.Packs == nil && options.Assets == ""
	singleton.RUnlock()
	if same {
		return nil
	}
	instance, err := New(options)
	if err != nil {
		return err
	}
	singleton.Lock()
	previous := current
	current = instance
	currentKey = key
	singleton.Unlock()
	if previous != nil {
		previous.Close()
	}
	return nil
}

// Judge scores one message with the package judge. Before Init and after
// Close it returns the fail-open verdict. It never fails.
func Judge(text string) Judgement {
	singleton.RLock()
	instance := current
	singleton.RUnlock()
	if instance == nil {
		return failOpen()
	}
	return instance.Judge(text)
}

// Ready reports whether Init has installed a judge that Close has not released.
func Ready() bool {
	singleton.RLock()
	defer singleton.RUnlock()
	return current != nil
}

// Close releases the package judge. Judge fails open until the next Init.
func Close() {
	singleton.Lock()
	previous := current
	current = nil
	currentKey = ""
	singleton.Unlock()
	if previous != nil {
		previous.Close()
	}
}

File: packages/javascript-common/src/singleton.ts
import type { Judge, InitOptions, Judgement } from "./contract.js";

/** The verdict for text nothing judges: the nudge fails open. */
export function failOpen(): Judgement {
  return { safe: true, score: 0, locale: null, grawlix: null };
}

/** One judge per module: `init` once, then call `judge` on every keystroke. */
export interface Singleton {
  /**
   * Loads the locales and installs the judge. Same options again reuse the
   * judge; different options build a new one and retire the old one only
   * after the new one is ready, so `judge` never has a gap. Rejects with the
   * coded errors of `createJudge` and leaves the previous judge in place.
   */
  init(options?: InitOptions): Promise<void>;
  /** Synchronous. Before `init` resolves, or after `close`, returns the fail-open verdict. Never throws. */
  judge(text: string): Judgement;
  /** True once `init` has resolved and until `close`. */
  ready(): boolean;
  /** Releases the packs. `judge` fails open until the next `init`. */
  close(): void;
}

function keyOf(options: InitOptions): string {
  const locales = Array.isArray(options.locales) ? [...options.locales].map(String).sort() : options.locales;
  return JSON.stringify({ locales, assets: options.assets ?? null, detectLanguage: options.detectLanguage ?? null, grawlix: options.grawlix ?? false });
}

/** Wraps a runtime's `createJudge` in module state. Every function is a closure, so callers may destructure them. */
export function createSingleton(create: (options: InitOptions) => Promise<Judge>): Singleton {
  let current: Judge | null = null;
  let currentKey: string | null = null;
  let pending: { key: string; promise: Promise<void> } | null = null;

  const init = (options: InitOptions = {}): Promise<void> => {
    const key = keyOf(options);
    if (current !== null && currentKey === key) {
      pending = null;
      return Promise.resolve();
    }
    if (pending !== null && pending.key === key) return pending.promise;
    const promise: Promise<void> = Promise.resolve().then(() => create(options)).then(
      (judge) => {
        // A newer init or a close superseded this one.
        if (pending?.promise !== promise) {
          judge.close();
          return;
        }
        const previous = current;
        current = judge;
        currentKey = key;
        pending = null;
        previous?.close();
      },
      (error: unknown) => {
        if (pending?.promise === promise) pending = null;
        throw error;
      },
    );
    pending = { key, promise };
    return promise;
  };

  const judge = (text: string): Judgement => (current === null ? failOpen() : current.judge(text));
  const ready = (): boolean => current !== null;
  const close = (): void => {
    pending = null;
    const previous = current;
    current = null;
    currentKey = null;
    previous?.close();
  };

  return { init, judge, ready, close };
}
